[Kwan Hao Wei] iP

KwanHW opened the pull request:
# Waldo

> Can't find me if I'm on Quarantine Order!

Waldo frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1.  download it from [here](https://github.com/KwanHW/ip/releases).
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x]     Managing tasks
- [x]     Managing deadlines
- [ ]     Reminders (coming soon)

<hr>
If you Java programmer, you can use it to practice Java too. Here's the main method:

```java
   public static void main(String[] args) {
        new Duke();
        Application.launch(Duke.class, args);
    }
```

KwanHW committed:
Implemented Level-1: Greeting, Echo and Exit

KwanHW committed:
Implemented Level-2: Add,List

KwanHW committed:
Implemented Level-3: Mark as Done

KwanHW committed:
Implemented Level-4: ToDos,Events, Deadlines

KwanHW committed:
Implemented basic TextUiTesting

KwanHW committed:
Refactored input processing and initialisation

KwanHW committed:
Implemented Level-5: Handle Errors

KwanHW committed:
Implemented Level-6:Delete

KwanHW committed:
Added enums to tasks

KwanHW committed:
Implemented Level-7: Save

KwanHW committed:
Merge pull request #1 from KwanHW/branch-Level-7

Branch level 7

KwanHW committed:
Implemented Level-8: Dates and Times

KwanHW committed:
Added support for tasks on a certain day

KwanHW committed:
Merge pull request #2 from KwanHW/branch-Level-8

Branch level 8

KwanHW committed:
Extracted codes into smaller components

KwanHW committed:
Abstracted more messages and minor ui fixes

KwanHW committed:
Organised classes into packages

KwanHW committed:
Add JUnit Testing to some of the classes

KwanHW committed:
Create JAR file

KwanHW committed:
Merge pull request #3 from KwanHW/refactor

Refactoring code

KwanHW committed:
Implement Level-9: Find

KwanHW committed:
Merge pull request #4 from KwanHW/branch-Level-9

Implement Level-9: Find

KwanHW committed:
Add JavaDoc to the classes and fix access visibility

KwanHW committed:
Add more documentation for remaining classes.

KwanHW committed:
Merge pull request #5 from KwanHW/branch-A-JavaDoc

Branch A-JavaDoc

KwanHW committed:
Implement A-CodingStandard

KwanHW committed:
Merge pull request #6 from KwanHW/branch-A-CodingStandard

Implement A-CodingStandard

LapisRaider commented on src/test/java/duke/task/TaskStoreTest.java:
```diff
@@ -0,0 +1,44 @@
+        LocalDate date = LocalDate.parse("2022-01-31");
+        ts.addTask(new Deadline("Complete Assignment", date));
+        ts.addTask(new Event("Project Meeting", date));
+        assertEquals("You don't have any tasks on Jan 29 2022", ts.getTasksOn(LocalDate.parse("2022-01-29")), "Failed: TaskStore contains tasks that do not fall on that date");
```
Line length longer than 120 characters, might want to warp it.

LapisRaider replied:
Line length longer than 120 characters, might want to wrap it.

LapisRaider commented on src/test/java/duke/task/TaskStoreTest.java:
```diff
@@ -0,0 +1,44 @@
+        LocalDate date = LocalDate.parse("2022-01-31");
+        ts.addTask(new Todo("Complete Assignment"));
+        ts.addTask(new Todo("Project Meeting"));
+        assertEquals("You don't have any tasks on Jan 29 2022", ts.getTasksOn(LocalDate.parse("2022-01-29")), "Failed: TaskStore only contains non-timeable instances.");
```
Line length longer than 120 characters, might want to wrap it.

LapisRaider commented on src/test/java/duke/task/TaskStoreTest.java:
```diff
@@ -0,0 +1,44 @@
+        LocalDate date = LocalDate.parse("2022-01-29");
+        ts.addTask(new Deadline("Complete Assignment", date));
+        ts.addTask(new Event("Project Meeting", date));
+        String expected = "Here are your tasks on Jan 29 2022\n" + ts.getTask(0).toString() + "\n" + ts.getTask(1).toString() + "\n";
```
Line length longer than 120 characters, might want to wrap it. There might be other lines longer than 120 characters that I couldn't catch. 

IntelliJ has this formatting option that will help you wrap the lines longer than a certain amount of characters (you can set the limit in settings). They have a bunch of other settings too, definitely a time saver instead of manually scanning through every line of the codes!

LapisRaider commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public boolean getIsDone() {
+        return this.isDone;
+    }
+
+    public String getDescription() {
+        return this.description;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
+
+    public void markAsUndone() {
+        this.isDone = false;
+    }
+
+    /**
+     * Generates a string which represents the task for disk storage. The format is as follows:
+     * <br>
+     * TaskType | isDone | Task Description
+     * <br>
+     * Note that isDone will be "1" if its is marked as done and "0" otherwise.
+     *
+     * @return String representation of the task for disk storage.
+     */
+    public String writeToFile() {
+        String isDone = this.isDone ? "1" : "0";
+        return String.join(" | ", this.taskType.toString(), isDone, this.description);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[%s][%s] %s", this.taskType.toString(), this.getStatusIcon(), this.description);
+    }
```
The usage of `this.` keyword. It is recommended to use the `this` keyword only when a field is shadowed by a method or constructor parameter.

![image](https://user-images.githubusercontent.com/36361212/152055483-4ee7ffc8-1a26-4e76-b477-275780702dbf.png)

LapisRaider commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,55 @@
+ * Represents an event that is happening on a certain date.
+ */
+public class Event extends Task implements Timeable {
+    protected LocalDate at;
```
Might want to change this to private instead of protected. For java, unlike C++ where protected makes the variable inaccessible to other classes besides the derived classes, in java, as long as a class is in the same package, they can access the protected variable.

If that is not your intention, you may want to change to private since Event class is only using this variable (correct me if I'm wrong).

KwanHW replied:
Thank you for catching that error. I may have gotten it mixed up with experience in previous languages (like C#). 

I have resolved the other issues similar as the one mentioned

LapisRaider commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+ * Represents a task that has to be completed before a certain date.
+ */
+public class Deadline extends Task implements Timeable {
+    protected LocalDate by;
```
Might want to change this to private instead of protected. For java, unlike C++ where protected makes the variable inaccessible to other classes besides the derived classes, in java, as long as a class is in the same package, they can access the protected variable.

If that is not your intention, you may want to change to private since Deadline class is only using this variable (correct me if I'm wrong).

LapisRaider commented on src/main/java/duke/io/Storage.java:
```diff
@@ -0,0 +1,90 @@
+package duke.io;
+
+import duke.task.*;
```
Might want to list out the imports explicitly instead of using wildcard imports since it's not recommended.

LapisRaider reviewed (commented):
Looks great! Just a few nits to fix.

TLDR:
- `this` keyword is recommended being used only when a field is shadowed by a method or constructor parameter.
- Some of your lines might be longer than 120 characters, IntelliJ can help you format.
- Protected might want to change to private. Protected keyword in Java makes the variable/function accessible for any class in the same package.
- Might want to change wildcard imports to listing explicitly instead.

KwanHW committed:
Add gradle support (#7)

* Configured Gradle for building Duke and automatically perform unit tests

KwanHW committed:
Added JavaFX and basic functionality

KwanHW committed:
Refactor and cleanup codebase

KwanHW committed:
Merge pull request #8 from KwanHW/branch-Level-10

Branch level 10

KwanHW committed:
Implement Checkstyle

KwanHW committed:
Merge pull request #9 from KwanHW/checkstyle-support

Implement Checkstyle

KwanHW committed:
Fix build.gradle file to build JavaFX

KwanHW committed:
Add assertions to the codebase

Currently, some classes in the codebase is not enforced with assertions which could lead to programmer-related errors.

The changes implemented will enforce certain assumptions and constraints to prevent such mistakes. (Example: duke.txt must exist when the program is initialised)

KwanHW committed:
Merge pull request #10 from KwanHW/branch-A-Assertions

Implement Assertions

KwanHW committed:
Added code readability improvements

There are several methods that require some readability improvements. Furthermore, some constants that belong to a class
were not organised to the rightful class that needs it.

By refactoring moving some of these values into proper constants, it is easier to make future changes to these constants while giving
meaningful names to better understand how they are derived within a function.

KwanHW committed:
Merge branch 'master' into branch-A-CodeQuality

KwanHW committed:
Merge pull request #11 from KwanHW/branch-A-CodeQuality

Improve code quality

KwanHW committed:
Implement check for clashing dates for new tasks

KwanHW committed:
Fix test cases due to adjusted task adding

KwanHW committed:
Convert Task into an abstract class

KwanHW committed:
Adjust implementation of delete,mark and unmark for mass op

KwanHW committed:
Extend MassOps to include 'all' keyword

KwanHW committed:
Add JavaDoc to updated functions

KwanHW committed:
Merge pull request #12 from KwanHW/extensions

Extensions

KwanHW committed:
Add CI Support

KwanHW committed:
Rename workflow directory

KwanHW committed:
Remove tracking of old workflow

KwanHW committed:
Update gradle.yml

KwanHW committed:
Fix CI (#13)

* Fix checkstyle error

* Fix test regression

* Removed text ui test in yml

KwanHW committed:
Fix code quality (#15)

* Removed dead code

* Formatted JavaDoc

* Remove excess returns statements

KwanHW committed:
Better GUI (#16)

* Adjust DialogBox to have more lines

* Change profile image of Waldo and User

* Improve the look of the UI

* Finalise user interface

KwanHW committed:
Amend build configuration

KwanHW committed:
Add first draft of user guide

KwanHW committed:
Allow case insensitive searching

KwanHW committed:
Fix unused imports

KwanHW committed:
Merge pull request #17 from KwanHW/fix-find-command

Allow case insensitive searching

KwanHW committed:
Finish first draft of UG

KwanHW committed:
Merge pull request #18 from KwanHW/userguide

KwanHW committed:
Fix commands not executing when empty
